##__1__. Подберите скорость обучения (alpha) и количество итераций (до совпадения ответов с результатами МНК):

In [2]:
import numpy as np

X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],   # для умножения на intercept
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]]) # стаж репетитора

y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60] # средний балл ЕГЭ (целевая переменная)

In [3]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

def calc_mae(y, y_pred):
    err = np.mean(np.abs(y - y_pred))
    return err

In [4]:
def model(X, alpha, weights, iter_count):
  n = X.shape[1]
  a = alpha
  w = np.array(weights)
  y_pred = 0
  w_needed = np.array([47.23214286,  3.91071429])
  for i in range(iter_count):
      y_pred = np.dot(w, X)
      err = calc_mse(y, y_pred)          
      for j in range(w.shape[0]):             
        w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))
      if np.allclose(w, w_needed,rtol=1e-11) and err == 45.937499999999986:
          print(i, w, err)
          # МНК: array([47.23214286,  3.91071429]) 45.937499999999986  

## В коде выше возникла проблема со сравнением массивов (точность после запятой указана, но np.allclose всё равно возвращает True при 47.23214285. Другие методы сравнения вообще не работают. Пришлось глазами искать значение. Буду признателен, если подскажете: как сравнивать такие массивы?

In [5]:
print(model(X, 0.013000000000000001, [1, 0.5], 1755))

1660 [47.23214285  3.91071429] 45.937499999999986
1678 [47.23214285  3.91071429] 45.937499999999986
1688 [47.23214285  3.91071429] 45.937499999999986
1690 [47.23214285  3.91071429] 45.937499999999986
1701 [47.23214285  3.91071429] 45.937499999999986
1707 [47.23214285  3.91071429] 45.937499999999986
1709 [47.23214285  3.91071429] 45.937499999999986
1712 [47.23214285  3.91071429] 45.937499999999986
1713 [47.23214285  3.91071429] 45.937499999999986
1724 [47.23214285  3.91071429] 45.937499999999986
1728 [47.23214285  3.91071429] 45.937499999999986
1729 [47.23214285  3.91071429] 45.937499999999986
1731 [47.23214285  3.91071429] 45.937499999999986
1732 [47.23214285  3.91071429] 45.937499999999986
1740 [47.23214285  3.91071429] 45.937499999999986
1749 [47.23214286  3.91071429] 45.937499999999986
None


## Alpha подобрана тем же методом перебора.
## Ответ: alpha = 0.013000000000000001, iter = 1749

# *2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [14]:
w = np.array([1, 0.5])
alpha = 0.01
n = X.shape[1]

for i in range(1000):
    y_pred = np.dot(w, X)
    err = calc_mse(y, y_pred)
    '''for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))'''
    w -= alpha * (1/n * 2 * (X @ (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # [47.23214286  3.91071429] 45.937499999999986

0 [2.102 3.9  ] 3173.15
100 [31.88770806  6.74418155] 175.19445858001842
200 [41.83683774  4.90699865] 61.9177717428135
300 [45.33508261  4.26102097] 47.913169919666785
400 [46.56511152  4.03388672] 46.181755648107604
500 [46.99760587  3.95402334] 45.96769776787538
600 [47.14967657  3.92594232] 45.941233404700036
700 [47.20314662  3.91606866] 45.93796156758051
800 [47.2219474   3.91259695] 45.93755706443538
900 [47.228558    3.91137626] 45.937507054979434


##*3. Вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию начинает изменяться ниже определенного порога. 

Сколько нужно сделать итераций, если установить допустимое отклонение mse в размере $\text{diff}=10^{-6}$, а значение $\alpha=10^{-2}$?

In [30]:
w = np.array([1, 0.5])
alpha = 1e-2
n = X.shape[1]
diff = 1
old_err = np.inf
new_err = np.inf
i=0
 
while diff >= 1e-6:
    y_pred = np.dot(w, X)
    new_err = calc_mse(y, y_pred)
    w -= alpha * (1/n * 2 * (X @ (y_pred - y)))
    i +=1
    diff = old_err - new_err
    old_err = new_err
print(i, w, new_err, diff)
     # [47.23214286  3.91071429] 45.937499999999986

810 [47.22286276  3.91242793] 45.937547277798465 9.987200257910445e-07
